src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1065,21 +1062,26 @@ private void onDecodeSubscribe(
                     final long newReplyId = supplyReplyId.applyAsLong(newInitialId);
                     final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
-                    final MqttSubscribeStream subscribeStream = new MqttSubscribeStream(newTarget,
-                                                    newRouteId, newInitialId, newReplyId, packetId, topicFilter, subscription);
-
-                    subscribeStream.doApplicationBegin(traceId, authorization, affinity, topicFilter, subscriptionId);
-
-                    correlations.put(newReplyId, subscribeStream);
-
                     if (topicFilter == null)
                     {
                         onDecodeError(traceId, authorization, PROTOCOL_ERROR);
                         break;
                     }
 
-                    final int key = topicKey(topicFilter);
-                    subscribers.put(key, subscribeStream);
+                    final int topicKey = topicKey(topicFilter);
+
+                    int finalSubscriptionId = subscriptionId;
+                    MqttServerStream stream = streams.computeIfAbsent(topicKey, s ->
+                    {
+                        final MqttServerStream newStream = new MqttServerStream(newTarget,
+                            newRouteId, newInitialId, newReplyId, packetId, topicFilter, subscription,
+                            MqttRole.RECEIVER);
+                        newStream.doApplicationBegin(traceId, authorization, affinity, topicFilter, finalSubscriptionId);
+                        return newStream;
+                    });
```

Comment:
Same feedback as above regarding side effects during `computeIfAbsent` and using `doApplicationBeginIfNecessary` instead.